### DISCUSSION TITLE
Add new setter method to TestClient for scope["client"]

### DISCUSSION THREAD

**Discussion by pythonweb2** [encode/starlette#2484] [Q&A]
Related to #2430.

@Kludex @aminalaee since this was rejected, can a new setter method be added to the class? I can write the code/tests of course, just would need a sign off. Really would be nice to be able to change this value in the test suite to test different scenarios.

e.g.
```python
class TestClient(httpx.Client):
    __test__ = False
    task: Future[None]
    portal: anyio.abc.BlockingPortal | None = None

    def __init__(
        self,
        app: ASGIApp,
        base_url: str = "http://testserver",
        raise_server_exceptions: bool = True,
        root_path: str = "",
        backend: typing.Literal["asyncio", "trio"] = "asyncio",
        backend_options: dict[str, typing.Any] | None = None,
        cookies: httpx._types.CookieTypes | None = None,
        headers: dict[str, str] | None = None,
        follow_redirects: bool = True,
    ) -> None:
        self._scope_client: typing.Optional[typing.Tuple[str, int]] = None
        ...

    @property
    def scope_client(self):
        return self._scope_client

    @scope_client.setter
    def scope_client(self, value: typing.Optional[typing.Tuple[str, int]]):
        # Add some logic here to set the value
        ...
```

**Comment by acdha** [comment]
I was just working around the same regression introduced by #2377 when I saw #2430. The change made it harder to test applications which do any sort of modification to the client value (in my case, allowing certain CDN headers to override the IP address) since there isn't an easy hook for [overriding the `client=None` assignment](https://github.com/encode/starlette/blob/ee3cdfd5dc93658a2bb45d5a9bb8387b867cc585/starlette/testclient.py#L310C9-L310C14) and since that's an accessor property which returns a new instance on every call it's non-trivial to mock middleware which modifies it.

I ended up with an approach like this in my FastAPI app:

```python
from fastapi import APIRouter, FastAPI, Request
from fastapi.responses import JSONResponse, PlainTextResponse
from fastapi.testclient import TestClient
from starlette.middleware.base import BaseHTTPMiddleware

from my_app.middleware import use_custom_ip_header

app = FastAPI()

router = APIRouter()


@router.get("/")
def client_address_echo(request: Request) -> PlainTextResponse:
    return JSONResponse({"client": request.client}, status_code=200)


app.include_router(router)
app.add_middleware(BaseHTTPMiddleware, dispatch=use_custom_ip_header)

client = TestClient(app)


def test_ip_header_override_on_errors():
    response = client.get(
        "/",
        headers={
            "Content-Type": "application/json",
            "Connecting-IP": "9.8.7.6",
            "X-Forwarded-For": "1.2.3.4",
        },
    )
    assert response.json() == {"client": ["9.8.7.6", 50000]}
```

That works but there are two things I didn't like about it. One is that it means this test – unlike everything else in the test suite – uses a custom application rather than the one which actually runs. The other is that I now have a block of code in my main application code which looks like this, which seems undesirable:

```python
    if not request.client:
        if request.scope["server"][0] != "testserver":
            raise SystemExit("This service is not correctly configured!")
        else:
            request.scope["client"] = Address("testclient", 50000)
```

Obviously that's not the end of the world but it suggests that there should be a way to populate the test client with those parameters so the production application code wouldn't need to include a workaround for this new quirk.
